libs/checkpoint-postgres/langgraph/checkpoint/postgres/__init__.py

```diff
@@ -400,4 +403,59 @@ def _cursor(self, *, pipeline: bool = False) -> Iterator[Cursor[DictRow]]:
                     yield cur
 
 
-__all__ = ["PostgresSaver", "BasePostgresSaver", "ShallowPostgresSaver", "Conn"]
+
+class ResilientPostgresSaver(PostgresSaver):
+    def __init__(
+        self,
+        conn: Conn,
+        pipe: Optional[Pipeline] = None,
+        serde: Optional[SerializerProtocol] = None,
+        max_retries: int = 3,
+        retry_delay: float = 2.0,  # seconds
+    ) -> None:
+        super().__init__(conn, pipe, serde)
+        self.max_retries = max_retries
+        self.retry_delay = retry_delay
+
+    def _execute_with_retries(self, query_func, *args, **kwargs):
+        retries = 0
+        while retries < self.max_retries:
+            try:
+                return query_func(*args, **kwargs)
+            except (OperationalError, ConnectionError) as e:
+                retries += 1
+                if retries >= self.max_retries:
+                    raise
+                time.sleep(self.retry_delay)
+                self._reconnect()
+
+    def _reconnect(self):
+        if hasattr(self.conn, "close"):
+            self.conn.close()
+        if isinstance(self.conn, ConnectionPool):
+            self.conn.close()
+            self.conn = ConnectionPool(self.conn.conninfo)  # Reinitialize connection pool
```

Comment:
we shouldn't re-initialize the entire connection pool, we should just let it acquire a new connection from the existing pool, in this branch you can just pass

Reply:
I guess we cannot do that because we are already closing the connection. Let me know if you have anything in mind.

Reply:
Is there anyway to do this directly at the connection / connection pool layer?